man/rc.7
.TH RC 7

.SH NAME
rc \- RTags client

.SH SYNOPSIS
.B rc [options]

.B rc [\-\-help|\-h]

.SH DESCRIPTION
RTags client.

.SH OPTIONS

.SS RDM options

.TP
.BR \-\-quit\-rdm|\-q\ [\fIarg\fR]
Tell server to shut down. If \fIarg\fR is broadcast:10 tell all rdm on the farm to quit with exit code 10.

.TP
.BR \-\-connect\-timeout\ \fItimeout\fR
Timeout for connecting to rdm in ms (default 1000).

.SS Project management

.TP
.BR \-\-clear|\-C
Clear projects.

.TP
.BR \-\-project|\-w\ [\fIname\fR]
If specified, select project matching \fIname\fR if unique, otherwise list all projects.

.TP
.BR \-\-delete\-project|\-W\ [\fIname\fR]
Delete all projects matching regexp.

.TP
.BR \-\-unload|\-u\ [\fIarg\fR]
Unload project(s) matching argument.

.TP
.BR \-\-reload\-projects|\-z
Reload projects from projects file.

.TP
.BR \-\-jobcount|\-j\ [\fIn\fR]
Set or query current job count.

.TP
.BR \-\-syncproject
Synchronize current project ASAP.


.SS Indexing

.TP
.BR \-\-compile|\-c\ [\fIargs\fR]
Pass compilation arguments to rdm.

.TP
.BR \-\-load\-compilation\-database|\-J\ [\fIjson\fR]
Load compile_commands.json from directory

.TP
.BR \-\-suspend\-file|\-X\ [\fIarg\fR]
Dump suspended files (don't track changes in these files) with no arg. Otherwise toggle suspension for arg.


.SS Querying

.TP
.BR \-\-follow\-location|\-f\ \fIlocation\fR
Follow this location.

.TP
.BR \-\-references\-name|\-R\ \fIarg\fR
Find references matching arg.

.TP
.BR \-\-references|\-r\ \fIlocation\fR
Find references matching this location.

.TP
.BR \-\-list\-symbols|\-S\ [\fIpattern\fR]
List symbol names matching \fIpattern\fR.

.TP
.BR \-\-find\-symbols|\-F\ \fIarg\fR
Find symbols matching arg.

.TP
.BR \-\-cursor\-info|\-U\ \fIlocation\fR
Get cursor info for this location.

.TP
.BR \-\-status|\-s\ [\fIarg\fR]                     Dump status of rdm. Arg can be symbols or symbolNames.

.TP
.BR \-\-is\-indexed|\-T\ [\fIfile\fR]
Check if rtags knows about, and is ready to return information about, this source file.

.TP
.BR \-\-is\-indexing
Check if rtags is currently indexing files.

.TP
.BR \-\-has\-filemanager\ [\fIoptional\fR]
Check if rtags has info about files in this directory.

.TP
.BR \-\-preprocess|\-E\ [\fIfile\fR]
Preprocess file.

.TP
.BR \-\-reindex|\-V\ [\fIpattern\fR]
Reindex all files or all files matching \fIpattern\fR.

.TP
.BR \-\-path|\-P\ [\fIpattern\fR]
Print files matching pattern.

.TP
.BR \-\-dump\-file|\-d\ \fIfile\fR
Dump source file.

.TP
.BR \-\-rdm\-log|\-g
Receive logs from rdm.

.TP
.BR \-\-fixits\ \fIfile\fR
Get fixits for file.

.TP
.BR \-\-remove|\-D\ \fIfile\fR
Remove file from project.

.TP
.BR \-\-find\-project\-root\ [\fIarg\fR]
Use to check behavior of find\-project\-root.

.TP
.BR \-\-find\-project\-build\-root\ [\fIarg\fR]
Use to check behavior of find\-project\-root for builds.

.TP
.BR \-\-sources\ \fIfile\fR
Dump sources for source file.

.TP
.BR \-\-dependencies\ \fIfile\fR
Dump dependencies for source file.

.TP
.BR \-\-reload\-file\-manager|\-B
Reload file manager.

.TP
.BR \-\-man
Output XML for xmltoman to generate man page for rc :\-)

.TP
.BR \-\-code\-complete\-at|\-l\ \fIlocation\fR
Code complete at location: arg is file:line:col.

.TP
.BR \-\-prepare\-code\-complete\-at|\-b\ \fIlocation\fR
Prepare code completion at location: arg is file:line:col.

.TP
.BR \-\-send\-diagnostics\ [\fIarg\fR]
Only for debugging. Send data to all \-g connections.

.TP
.BR \-\-dump\-completions
Dump cached completions.


.SS Common


.TP
.BR \-\-strip\-paren|\-p
Strip parens in various contexts.

.TP
.BR \-\-max|\-M\ \fIlines\fR
Max lines of output for queries.

.TP
.BR \-\-reverse\-sort|\-O
Sort output reversed.

.TP
.BR \-\-unsaved\-file\ \fIfile\fR
Pass unsaved file on command line. E.g. \-\-unsaved\-file=main.cpp:1200 then write 1200 bytes on stdin.

.TP
.BR \-\-log\-file|\-L\ \fIfile\fR
Log to this file.

.TP
.BR \-\-no\-context|\-N
Don't print context for locations.

.TP
.BR \-\-path\-filter|\-i\ \fIpattern\fR
Filter out results not matching \fIpattern\fR.

.TP
.BR \-\-range\-filter\ [\fIarg\fR]
Filter out results not in the specified range.

.TP
.BR \-\-filter\-system\-headers|\-H
Don't exempt system headers from path filters.

.TP
.BR \-\-all\-references|\-e
Include definitions/declarations/constructors/destructors for references. Used for rename symbol.

.TP
.BR \-\-elisp\-list|\-Y
Output elisp: (list "one" "two" ...).

.TP
.BR \-\-diagnostics|\-G
Receive continual diagnostics from rdm.

.TP
.BR \-\-xml\-diagnostics|\-m
Receive continual XML formatted diagnostics from rdm.

.TP
.BR \-\-match\-regexp|\-Z
Treat various text patterns as regexps (\-P, \-i, \-V).

.TP
.BR \-\-match\-icase|\-I
Match case insensitively

.TP
.BR \-\-absolute\-path|\-K
Print files with absolute path.

.TP
.BR \-\-socket\-file|\-n\ [\fIfile\fR]
Use this socket file (default ~/.rdm).

.TP
.BR \-\-timeout|\-y\ \fItimeout\fR
Max time in ms to wait for job to finish (default no timeout).

.TP
.BR \-\-find\-virtuals|\-k
Use in combinations with \-R or \-r to show other implementations of this function.

.TP
.BR \-\-find\-file\-prefer\-exact|\-A
Use to make \-\-find\-file prefer exact matches over partial matches.

.TP
.BR \-\-cursorinfo\-include\-parents
Use to make \-\-cursor\-info include parent cursors.

.TP
.BR \-\-cursorinfo\-include\-targets
Use to make \-\-cursor\-info include target cursors.

.TP
.BR \-\-cursorinfo\-include\-references
Use to make \-\-cursor\-info include reference cursors.

.TP
.BR \-\-cursor\-kind
Include cursor kind in \-\-find\-symbols output.

.TP
.BR \-\-display\-name
Include display name in \-\-find\-symbols output.

.TP
.BR \-\-current\-file\ \fIfile\fR
Pass along which file is being edited to give rdm a better chance at picking the right project.

.TP
.BR \-\-declaration\-only
Filter out definitions (unless inline).

.TP
.BR \-\-imenu
Use with \-\-list\-symbols to provide output for (rtags\-imenu) (filter namespaces, fully qualified function names, ignore certain cursors etc).

.TP
.BR \-\-containing\-function|\-o
Include name of containing function in output.

.TP
.BR \-\-build\-index\ [\fIarg\fR]
For sources with multiple builds, use the arg'th.

.TP
.BR \-\-compilation\-flags\-only
For \-\-source, only print compilation flags.

.TP
.BR \-\-compilation\-flags\-split\-line
For \-\-source, print one compilation flag per line.

.TP
.BR \-\-dump\-include\-headers
For \-\-dump\-file, also dump dependencies.

.TP
.BR \-\-silent\-query
Don't log this request in rdm.

.TP
.BR \-\-synchronous\-completions
Wait for completion results.

.TP
.BR \-\-unescape\-compile\-commands
Unescape \'s and unquote arguments to \-c.

.TP
.BR \-\-no\-unescape\-compile\-commands
Escape \'s and unquote arguments to \-c.

.TP
.BR \-\-no\-sort\-references\-by\-input
Don't sort references by input position.

.TP
.BR \-\-project\-root\ [\fIarg\fR]
Override project root for compile commands.

.TP
.BR \-\-rtags\-config\ [\fIarg\fR]
Print out .rtags\-config for argument.


.SH EXAMPLES
.SH ENVIRONMENT
.SH SEE ALSO



rc.c
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <limits.h>
#include <assert.h>
#include "Shared.h"
#include <getopt.h>

static int32_t locationLength = -1;

static int find(const void *l, const void *r)
{
    const char *left = ((const char*)l) + Int32Length;
    const char *right = ((const char*)r) + Int32Length;
    // printf("%s %s %s\n", left, right, std::string(left, locationLength).c_str());
    return strncmp(left, right, locationLength - Int32Length);
}


/* struct Cleanup { */
/*     ~Cleanup() */
/*     { */
/*         if (*mapped) */
/*             munmap(*mapped, st->st_size); */
/*         if (*fd) */
/*             close(*fd); */
/*     } */

/*     int *fd; */
/*     struct stat *st; */
/*     void **mapped; */
/* }; */

void recurse(const char *ch, int32_t pos, int indent)
{
    struct NodeData node = readNodeData(ch + pos);
    int i;
    for (i=0; i<indent; ++i) {
        printf(" ");
    }
    printf("%s %s %s\n", nodeTypeToName(node.type, Normal), node.symbolName,
           node.location ? ch + node.location : "");
    if (node.firstChild)
        recurse(ch, node.firstChild, indent + 2);
    if (node.nextSibling)
        recurse(ch, node.nextSibling, indent);
}

static inline void usage(FILE *f)
{
    fprintf(f,
            "rc [options]...\n"
            "  --help|-h                Display this help\n"
            "  --follow-symbol|-s [arg] Follow this symbol (e.g. /tmp/main.cpp:32:1)\n"
            "  --references|-r [arg]    Print references of symbol at arg\n"
            "  --print-tree|-t          Print out the node tree to stdout\n"
            "  --db-file|-f [arg]       Use this database file\n");
}

int main(int argc, char **argv)
{
    struct option longOptions[] = {
        { "help", 0, 0, 'h' },
        { "follow-symbol", 1, 0, 's' },
        { "print-tree", 0, 0, 't' },
        { "db-file", 1, 0, 'f' },
        { "references", 1, 0, 'r' },
        { 0, 0, 0, 0 },
    };
    const char *shortOptions = "hs:tf:r:";
    int idx;
    int longIndex;
    const char *arg = 0;
    const char *dbFile = 0;
    enum Mode {
        None,
        FollowSymbol,
        References
    } mode = None;
    char dbFileBuffer[PATH_MAX + 10];
    int showTree = 0;
    // for (int i=0; i<argc; ++i) {
    //     printf("%d %s\n", i, argv[i]);
    // }

    while ((idx = getopt_long(argc, argv, shortOptions, longOptions, &longIndex)) != -1) {
        switch (idx) {
        case '?':
            usage(stderr);
            return 1;
        case 'h':
            usage(stdout);
            return 0;
        case 's':
            arg = optarg;
            mode = FollowSymbol;
            break;
        case 'r':
            arg = optarg;
            mode = References;
            break;
        case 't':
            showTree = 1;
            break;
        case 'f':
            dbFile = optarg;
            break;
        }
    }

    if (!dbFile) {
        if (getcwd(dbFileBuffer, PATH_MAX)) {
            const int len = strlen(dbFileBuffer);
            if (len > 0 && dbFileBuffer[len - 1] != '/') {
                dbFileBuffer[len] = '/';
                dbFileBuffer[len + 1] = '\0';
            }
            char *slash;
            while ((slash = strrchr(dbFileBuffer, '/'))) {
                // ### this is awful
                strcpy(slash + 1, ".rtags.db");
                struct stat s;
                // printf("Testing [%s]\n", dbFileBuffer);
                if (stat(dbFileBuffer, &s) >= 0) {
                    dbFile = dbFileBuffer;
                    break;
                }
                *slash = '\0';
            }
        }
        if (!dbFile) {
            printf("%s %d: if (!dbFile) {\n", __FILE__, __LINE__);
            return 1;
        }
    }

    if (mode == None && !showTree) {
        printf("%s %d: if (mode == None && !showTree) {\n", __FILE__, __LINE__);
        return 1;
    }
        

    int fd = 0;
    struct stat st;
    void *mapped = 0;
    const char *ch = 0;

    fd = open(dbFile, O_RDONLY);
    if (fd <= 0) {
        printf("%s %d: if (fd <= 0)\n", __FILE__, __LINE__);
        return 1;
    }

    if (fstat(fd, &st) < 0) {
        printf("%s %d: if (fstat(fdin, &st) < 0) \n", __FILE__, __LINE__);
        close(fd);
        return 1;
    }
    if (st.st_size < 10) {
        printf("%s %d: if (st.st_size < 10) {\n", __FILE__, __LINE__);
        close(fd);
        return 1;
    }
        
    if ((mapped = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0)) == MAP_FAILED) {
        printf("%s %d: if ((src = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0)) == MAP_FAILED) {\n", __FILE__, __LINE__);
        close(fd);
        return 1;
    }
    ch = (char*)mapped;
    if (strncmp(mapped, "Rt", 3)) {
        printf("%s %d: if (memcmp(mapped, \"Rt\", 2)) {\n", __FILE__, __LINE__);
        munmap(mapped, st.st_size);
        close(fd);
        return 1;
    }

    // for (int i=0; i<10; ++i) {
    //     printf("%d %x %c\n", i, ch[i], ch[i]);
    // }

    const int32_t nodeCount = readInt32(ch + NodeCountPos);
    locationLength = readInt32(ch + IdLengthPos);
    const int32_t dictionaryPosition = readInt32(ch + DictionaryPosPos);
    const int32_t dictionaryCount = readInt32(ch + DictionaryCountPos);
    const int32_t dictionarySymbolLengthPos = readInt32(ch + DictionarySymbolNameLengthPos);
    const int32_t dictionaryMaxSynonyms = readInt32(ch + DictionaryMaxSynonymsPos);
    /* printf("%d %d - %d %d %d %d\n", locationLength, nodeCount, dictionaryPosition, dictionaryCount, dictionarySymbolLengthPos, dictionaryMaxSynonyms); */
    // qDebug() << (locationLength + 1 + Int32Length);
    if (locationLength <= 0 || nodeCount <= 0) {
        munmap(mapped, st.st_size);
        close(fd);
        printf("%s %d: if (locationLength <= 0 || nodeCount <= 0)\n", __FILE__, __LINE__);
        return 1;
    }

    if (showTree) {
        recurse(ch, rootNodePosition(nodeCount, locationLength), 0);
    }

    if (arg) {
        const int argLen = strlen(arg) + 1;
        char *padded = (char*)malloc(argLen + Int32Length);
        strncpy(padded + Int32Length, arg, argLen);
        const char *bs = (const char*)bsearch(padded, ch + FirstId, nodeCount, locationLength, find);
        // printf("Found a match %p\n", bs);
        free(padded);
        if (bs) {
            const int32_t idx = readInt32(bs);
            struct NodeData node = readNodeData(ch + idx);
            if (mode == References) {
                if (node.firstChild) {
                    node = readNodeData(ch + node.firstChild);
                    while (1) {
                        if (node.type == Reference && node.location) {
                            printf("%s\n", ch + node.location);
                        }
                        if (!node.nextSibling)
                            break;
                        node = readNodeData(ch + node.nextSibling);
                    }
                }
            } else {
                // printf("Found node %s %s\n", nodeTypeToName(type), symbolName);
                int32_t found = 0;
                NodeType targetType = MethodDeclaration;
                switch (node.type) {
                case MethodDeclaration:
                    targetType = MethodDefinition;
                case MethodDefinition: {
                    struct NodeData parent = readNodeData(ch + node.parent);
                    assert(parent.firstChild);
                    struct NodeData sibling = readNodeData(ch + parent.firstChild);
                    while (1) {
                        if (sibling.type == (int)targetType && !strcmp(node.symbolName, sibling.symbolName)) {
                            found = sibling.location;
                            break;
                        }
                        if (!sibling.nextSibling)
                            break;
                        sibling = readNodeData(ch + sibling.nextSibling);
                    }
                    break; }
                case Reference:
                case EnumValue:
                    found = readNodeData(ch + node.parent).location;
                    break;
                default:
                    break;
                }
                if (found) {
                    printf("%s\n", ch + found);
                } else {
                    printf("Couldn't find it\n");
                }
            }
        }
    }
    munmap(mapped, st.st_size);
    close(fd);
    return 0;
}

